### PR TITLE
Added support for optional HMACOutputLength for hmac-sha signing & va…

### PR DESCRIPTION
**Pull request by eclhmf**
…lidation.

### CODE REVIEW

**Comment by eclhmf** [comment]
This pull request adds support for the optional HMACOutputLength parameter, which truncates the HMAC output digest. See https://www.w3.org/TR/xmldsig-core1/#sec-SignatureMethod.

File: signxml/__init__.py [hunk 1]
```diff
@@ -896,7 +916,15 @@ def verify(self, data, require_x509=True, x509_cert=None, cert_subject_name=None
                           algorithm=self._get_hmac_digest_method(signature_alg),
                           backend=default_backend())
             signer.update(signed_info_c14n)
-            if raw_signature != signer.finalize():
+            verify_signature = signer.finalize()
+            if hmac_output_length is not None:
+                # Truncating the HMAC result, see https://www.w3.org/TR/xmldsig-core/#sec-SignatureMethod
+                result_length = int(hmac_output_length.text) // 8
+                if(result_length < max(10, signer.algorithm.digest_size // 2)):
+                    raise InvalidInput('HMAC output length is too small')
```
Comment (kislyuk) [inline signxml/__init__.py:924]: This should be `InvalidSignature`

**Review by kislyuk — CHANGES_REQUESTED** [review]
Thank you for your contribution.

Do you have a specific application that requires the use of this parameter? This parameter is a security hazard since it can be used in downgrade attacks. Moreover, your implementation introduces a downgrade attack vulnerability because it allows a malicious signer to downgrade ANY algorithm, not just the HMAC-SHA ones.

While I initially declined to implement `ds:HMACOutputLength` due to the security hazards associated with it, I can consider it as impelmented in this PR if the following issues are addressed:

* Document an interoperability rationale in the form of a library or other publicly distributed application that requires it
* Raise an `InvalidSignature` error if `ds:HMACOutputLength` appears with a signature method other than `HMAC-SHA*`
* Disable the use of `ds:HMACOutputLength` in `XMLVerifier.verify()` by default due to the security challenges it presents. Add a class variable or keyword argument, `allow_hmac_output_length`, set to False by default and configurable by the caller
* Add a test case asserting correct behaviors in all of the cases described above
* Enable all test vectors using `ds:HMACOutputLength` in the test suite and verify that they pass.

**Comment by kislyuk** [comment]
We will not be implementing HMAC truncation or any peer-configurable signature or digest material truncation due to the security hazards inherent in such functionality. See https://www.w3.org/blog/2009/07/hmac-truncation-in-xml-signatu/ for further context.

**Pull request closed without merging**